core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -362,24 +351,6 @@ class BrokerToControllerRequestThread(
     }
   }
 
-  override def doWork(): Unit = {
-    if (activeControllerAddress().isDefined) {
-      super.pollOnce(Long.MaxValue)
-    } else {
-      debug("Controller isn't cached, looking for local metadata changes")
-      controllerNodeProvider.get() match {
-        case Some(controllerNode) =>
-          info(s"Recorded new controller, from now on will use broker $controllerNode")
-          updateControllerAddress(controllerNode)
-          metadataUpdater.setNodes(Seq(controllerNode).asJava)
-        case None =>
```

Comment:
If there is no controller defined yet, this logic is still useful.

Reply:
It seems this should not happen at runtime, since the very 1st request received by a broker should be the UpdateMetadata request from a controller containing the controller node, which will populate the metadataCache.
After that, each subsequent UpdateMetadata request should always contain the controller node.
Hence, the controller node should always be available (although it may change).


In the updated PR, even if somehow there is no controller defined, then the NetworkClient will not see any requests-to-be-sent:
https://github.com/linkedin/kafka/blob/eb12c165f1cd3d251c8ce66db4aa614fdd5cec5a/core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala#L317

The NetworkClient will be blocking waiting for a max of 600s, 
https://github.com/linkedin/kafka/blob/eb12c165f1cd3d251c8ce66db4aa614fdd5cec5a/core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala#L354
and then query for the controller node again through the ControllerNodeProvider.